### DISCUSSION TITLE
DSL parser and compiler written in Rust

### DISCUSSION THREAD

**Discussion by iammathew** [openfga/community#51] [Show and tell]
Hey OpenFGA community,

just wanted to share this small project of mine here. Since the current DSL parser is a nodejs package and therefore not as easy-to-use depending on your use case, I built a rust version of it, the code is still very early stage (also my Rust experience is not very strong).

https://github.com/iammathew/openfga-rs

Right now, unfortunately, it does not have any correctness checking of the parsed model, which I plan to add next, other planned changes are to make a WebAssembly module available so it can run in the browser or wherever you need it!
If that is done, I'm thinking about actually using the same parser to create a simple LSP for better code navigation inside VSCode.

Also I feel like currently it may not include all features of the DSL, but can parse all of the published examples.

Feel free to let me know about your thoughts and opinions!

**Comment by rhamzeh** [comment]
Thanks for sharing this @iammatthew!

Linked to it in this [PR](https://github.com/openfga/syntax-transformer/pull/74), please take a look and let me know if I got everything right

In terms of community collaboration, I would love to have a base set of tests for all the projects - we can export them from the syntax-transformer repo for all of you to contribute to and consume.

I wonder what would be a good format for all of you.

A single json file?

A folder with multiple json files, each containing the JSON and the DSL syntax?

Multiple folder with one file for the JSON, one for the DSL?

We would need to export:
- Tests for parsing in schema v1.0
- Tests for parsing in schema v1.1
- Tests for checking schema v1.0
- Tests for checking schema v1.1

So maybe:
```
/tests/data/
- v1.0/
  - parser/
    - {case name}/
      - authorization-model.json
      - authorization-model.openfga
  - checker/
    - dsl/ 
      - {case name}/
        - authorization-model.openfga
        - results.json
    - json/ 
      - {case name}/
        - authorization-model.json
        - results.json
  - formatter/
    - {case name}/
      - authorization-model.input.openfga
      - authorization-model.output.openfga
- v1.1/
  - parser/
  - checker/
  - formatter/
```

(cc @adriantam)

**Reply by MidasLamb** [reply]
I'd argue for potentially moving from `/{case name}/authorization-model.json` &`/{case name}/authorization-model.openfga`  to `/{case name}.json` & `/{case name}.openfga`, this prevents a lot of files from having the same name, just being in a different directory. For the formatter and json the results could then probably be something along the lines of `/{case name}.results.json`

**Reply by iammathew** [reply]
Thanks for checking!
I don't have any strong preference regarding the structure both proposed by you (@rhamzeh) and @MidasLamb look good to me! Only remark as long as it is 2 files per test case having them in a flat folder seems okay, once it becomes more (which I hope they won't) I prefer the folder structure, just for navigating purposes.

**Reply by iammathew** [reply]
@rhamzeh Heyah! Is there any update on the combined test suite?

**Reply by rhamzeh** [reply]
Not yet! Work on that will probably start after [this PR](https://github.com/openfga/syntax-transformer/pull/75) gets merged and released so as not to increase scope for the next release & support for v1.1 schemas.

On the upside we have resolved [the freezing issue](https://github.com/openfga/syntax-transformer/issues/77) you were facing so that should also be out in the `v0.0.7` release.

Edit: Tracking this issue [here](https://github.com/openfga/syntax-transformer/issues/84)
